### PR TITLE
Use Sendgrid for email confirmation and password reset

### PR DESCRIPTION
**Pull request by joshstricklin**

### CODE REVIEW

**Comment by joshstricklin** [comment]
[ch34]
[ch35]

**Comment by vishalarora** [comment]
The deploy script failed because the seed tried to send an email.  You might need to make an exception for that.

File: .gitignore [hunk 1]
```diff
@@ -11,4 +11,5 @@
 /public/assets
 /tags
 /tmp/*
-/node_modules
\ No newline at end of file
+/node_modules
+/sendgrid.env
```
Comment (vishalarora) [inline .gitignore]: Are we actually using a separate ENV file for Sendgrid? Can we just use `.env`?

File: config/environment.rb [hunk 2]
```diff
@@ -1,2 +1,12 @@
 require_relative 'application'
 Rails.application.initialize!
+
+ActionMailer::Base.smtp_settings = {
+  :user_name => ENV['SENDGRID_USERNAME'],
+  :password => ENV['SENDGRID_PASSWORD'],
+  :domain => 'joshstricklin.com',
```
Comment (vishalarora) [inline config/environment.rb]: Can we read this from an environment variable?

File: config/environment.rb [hunk 3]
```diff
@@ -1,2 +1,12 @@
...
+  :address => 'smtp.sendgrid.net',
```
Comment (vishalarora) [inline config/environment.rb]: Can we read this from an environment variable?

File: config/initializers/devise.rb [hunk 4]
```diff
@@ -12,10 +12,10 @@
   # Configure the e-mail address which will be shown in Devise::Mailer,
   # note that it will be overwritten if you use your own mailer class
   # with default "from" parameter.
-  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
+  config.mailer_sender = 'taskopolis@joshstricklin.com'
```
Comment (vishalarora) [inline config/initializers/devise.rb]: Can we read this from an environment variable?

**Review by vishalarora — COMMENTED** [review]
The deploy script failed because the seed tried to send an email. You might need to make an exception for that.

File: config/initializers/devise.rb (hunk 4 above)
Comment (joshstricklin) [reply config/initializers/devise.rb]: Good call.

File: .gitignore (hunk 1 above)
Comment (joshstricklin) [reply .gitignore]: That was for my local environment - ended up going a different route anyway, so I'll remove it.

File: app/mailers/application_mailer.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: 'taskopolis@joshstricklin.com'
```
Comment (vishalarora) [inline app/mailers/application_mailer.rb]: Can we read this from an environment variable?

File: app.json [hunk 6]
```diff
@@ -33,6 +33,24 @@
     },
     "SMTP_USERNAME":{
```
Comment (vishalarora) [inline app.json]: Do we need these other mail-related ENV variables since Sendgrid is installed now?

File: app.json [hunk 7]
```diff
@@ -33,6 +33,24 @@
...
+    "SENDGRID_API_KEY":{
+      "required":true
+    },
+    "MAILER_DOMAIN":{
+      "required":true
+    },
+    "MAILER_ADDRESS":{
+      "required":true
+    },
+    "MAILER_SENDER":{
```
Comment (vishalarora) [inline app.json]: Can these names be a bit more clear?  Ideally, we shouldn't have to refer to the code to figure out what these are.

Suggestion `DEFAULT_SEND_FROM_ADDRESS`

File: app/models/user.rb [hunk 8]
```diff
@@ -2,7 +2,8 @@ class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :trackable, :validatable
+         :recoverable, :rememberable, :trackable, :validatable,
+         :confirmable
```
Comment (vishalarora) [inline app/models/user.rb]: I think this is from a previous PR so you'll need to rebase :)

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 <p>Welcome <%= @email %>!</p>
 
+<p>Thanks for signing up for Taskopolis.</p>
```
Comment (vishalarora) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: No, thank you!

File: db/migrate/20170626182111_add_confirmable_to_devise.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+class AddConfirmableToDevise < ActiveRecord::Migration[5.0]
```
Comment (vishalarora) [inline db/migrate/20170626182111_add_confirmable_to_devise.rb]: Rebase should remove this, I think.

File: db/schema.rb [hunk 11]
```diff
@@ -10,7 +10,7 @@
 #
```
Comment (vishalarora) [inline db/schema.rb]: Same rebase comment

**Review by vishalarora — APPROVED** [review]
👍 Some comments about naming but this looks good to me.

Make sure you've added an API key to the prod and staging environments.

File: db/migrate/20170626182111_add_confirmable_to_devise.rb (hunk 10 above)
Comment (vishalarora) [inline db/migrate/20170626182111_add_confirmable_to_devise.rb]: Shut up, @vishalarora

File: db/schema.rb (hunk 11 above)
Comment (vishalarora) [inline db/schema.rb]: LIAR!

File: app/models/user.rb (hunk 8 above)
Comment (vishalarora) [inline app/models/user.rb]: ☝️  Idiotic comment

File: app.json (hunk 7 above)
Comment (joshstricklin) [reply app.json]: 👍

File: app/views/devise/mailer/confirmation_instructions.html.erb (hunk 9 above)
Comment (joshstricklin) [reply app/views/devise/mailer/confirmation_instructions.html.erb]: ❤️

File: app.json (hunk 6 above)
Comment (joshstricklin) [reply app.json]: Good call, I'll get rid of anything unnecessary.

File: config/smtp.rb [hunk 12]
```diff
@@ -1,11 +1,11 @@
 SMTP_SETTINGS = {
```
Comment (vishalarora) [inline config/smtp.rb]: Is this required if you have `ActionMailer::Base.smtp_settings` in `config/environment.rb`?

I have a feeling that it's redundant.  So, you might want to either delete `smtp.rb` altogether or move the config from the environment config into `smtp.rb`

File: db/schema.rb [hunk 13]
```diff
@@ -56,12 +55,12 @@
     t.inet     "last_sign_in_ip"
     t.datetime "created_at",                             null: false
     t.datetime "updated_at",                             null: false
-    t.integer  "points"
     t.string   "first_name"
     t.string   "last_name"
     t.string   "confirmation_token"
     t.datetime "confirmed_at"
     t.datetime "confirmation_sent_at"
+    t.integer  "points",                 default: 0
```
Comment (vishalarora) [inline db/schema.rb:63]: This seems like a change I made and should already be in `master`.  It's not destructive though so I don't think we need to spend any time debugging it.

**Pull request merged**